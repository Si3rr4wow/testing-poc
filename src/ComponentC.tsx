import { ChangeEventHandler, useState } from 'react'
import ComponentB from "./ComponentB"

const ComponentC: React.FC = () => {
  const [id, setId] = useState<string>('')

  const handleValueChange: ChangeEventHandler<HTMLInputElement> = ({ target: { value } }) => setId(value)

  return (
    <div className="component-wrapper">
      <code className="component-label">Component C</code>
      <div>
        <label htmlFor="id-input">Enter a person's ID: </label>
        <input id="id-input" value={id} onChange={handleValueChange} placeholder="abc123"/>
      </div>
      <ComponentB id={id}/>
    </div>
  )
}

export default ComponentC