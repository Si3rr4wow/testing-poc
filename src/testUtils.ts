export const getRandomString = (length = 8): string => 
  Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, length);
