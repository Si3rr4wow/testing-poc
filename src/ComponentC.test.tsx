import { fireEvent, render, RenderResult } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import ComponentC from './ComponentC'
import findPerson from './findPerson'
import { getRandomString } from './testUtils'
jest.mock('./findPerson', () => ({
  __esModule: true,
  default: jest.fn()
}))

describe('ComponentC', () => {
  describe('when I type in the input', () => {
    // Unit test for ComponentC
    it('displays what I type in the input', async () => {
      const id = getRandomString()
      const name = getRandomString();
      (findPerson as jest.Mock).mockImplementation(() => Promise.resolve({ name }))
      const { getByPlaceholderText } = render(<ComponentC/>)
      const input = getByPlaceholderText('abc123')
      await act(async () => {
        fireEvent.change(input, { target: { value: id } })
      })
      expect((input as HTMLInputElement).value).toBe(id)
    })
    // Integration test of C and B
    it('has a child that fetches the person with the id I typed', async () => {
      /* === BEGINNING OF C UNIT TEST ===*/
      const id = getRandomString()
      const name = getRandomString();
      (findPerson as jest.Mock).mockImplementation(() => Promise.resolve({ name }))
      const { getByPlaceholderText } = render(<ComponentC/>)
      const input = getByPlaceholderText('abc123')
      await act(async () => {
        fireEvent.change(input, { target: { value: id } })
      })
      // === END OF B UNIT TEST === //
      expect(findPerson).toHaveBeenCalledWith(id)
    })
  })
    // Integration test of C, B and A
    it('matches it`s snapshot', async () => {
      const name = 'Snapshot name';
      (findPerson as jest.Mock).mockImplementation(() => Promise.resolve({ name }))
      let renderResult = {} as RenderResult;
      await act(async () => {
        renderResult = render(<ComponentC/>)
      })
      expect(renderResult.asFragment()).toMatchSnapshot()
    })
})