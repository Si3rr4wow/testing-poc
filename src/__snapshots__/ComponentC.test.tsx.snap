// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentC matches it\`s snapshot 1`] = `
<DocumentFragment>
  <div
    class="component-wrapper"
  >
    <code
      class="component-label"
    >
      Component C
    </code>
    <div>
      <label
        for="id-input"
      >
        Enter a person's ID: 
      </label>
      <input
        id="id-input"
        placeholder="abc123"
        value=""
      />
    </div>
    <div
      class="component-wrapper"
    >
      <code
        class="component-label"
      >
        Component B
      </code>
      <p>
        finding person with id 
      </p>
      <div
        class="component-wrapper"
      >
        <code
          class="component-label"
        >
          Component A
        </code>
        <p>
          The person's name is Snapshot name
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
